constraints/miden-vm/hash.air

```diff
@@ -0,0 +1,76 @@
+mod HashAir
+
+### Constants and periodic columns ################################################################
+
+periodic_columns:
+    k0: [0, 0, 0, 0, 0, 0, 0, 1]
+    k1: [0, 0, 0, 0, 0, 0, 1, 0]
+    k2: [1, 0, 0, 0, 0, 0, 0, 0]
+
+### Helper functions/evaluators ###################################################################
+
+# Enforces that column must be binary.
+ev is_binary(main: [a]):
+    enf a^2 - a = 0
+
+### Bitwise Chiplet Air Constraints ###############################################################
+
+# Enforces the constraints on the hash chiplet, given the columns of the hash execution trace.
+ev hash_chiplet(main: [s[3], r, h[12], i]):
+
+    ## Instruction flags ##
+    let f_rpr = 1 - k0
+    let f_bp = k2 * s[0] * (1 - s[1]) * (1 - s[2])
+    let f_mp = k2 * s[0] * (1 - s[1]) * s[2]
+    let f_mv = k2 * s[0] * s[1] * (1 - s[2])
+    let f_mu = k2 * s[0] * s[1] * s[2]
+    let f_hout = k0 * (1 - s[0]) * (1 - s[1]) * (1 - s[2])
+    let f_sout = k0 * (1 - s[0]) * (1 - s[1]) * s[2]
+    let f_out = k0 * (1 - s[0]) * (1 - s[1])
+    let f_out_next = k1 * (1 - s[0]') * (1 - s[1]')
+    let f_abp = k0 * s[0] * (1 - s[1]) * (1 - s[2])
+    let f_mpa = k0 * s[0] * (1 - s[1]) * s[2]
+    let f_mva = k0 * s[0] * s[1] * (1 - s[2])
+    let f_mua = k0 * s[0] * s[1] * s[2]
+
+    ## Row address constraint ##
+    # TODO: Apply row address constraint
+
+    ## Selector columns constraints ##
+    # Enforce that selector columns are binary.
+    enf is_binary([s_i]) for s_i in s
+    # Enforce that unless f_out = 1 or f_out' = 1, the values in columns s[1] and s[2] are copied 
+    # over to the nex row.
+    enf (s[1]' - s[1]) * (1 - f_out_next) * (1 - f_out) = 0
+    enf (s[2]' - s[2]) * (1 - f_out_next) * (1 - f_out) = 0
```

Comment:
Since these are constraints rather than variables, can we make use of selectors to do the following?

```
enf s[1]' = s[1] when !f_out & !f_out_next
enf s[2]' = s[2] when !f_out & !f_out_next
```

In fact, it might be best to make an `is_copied` function, because we check for that a lot in this module and you could use it elsewhere as well


```
enf is_copied([s[1]]) when !f_out & !f_out_next
enf is_copied([s[2]]) when !f_out & !f_out_next
```